portfolio/src/main/java/com/google/sps/servlets/AuthServlet.java

```diff
@@ -51,6 +60,82 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
     response.getWriter().println(json);
   }
 
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    UserService userService = UserServiceFactory.getUserService();
+
+    // If the user is not logged in, but still resets their username, redirect to index.
+    if (!userService.isUserLoggedIn()) {
+      response.sendRedirect("/");
+      return;
+    }
+
+    // Add the username to the database
+    String username = getParameter(request, "text-input", "");
+    String id = userService.getCurrentUser().getUserId();
+    putUsername(username, id);
+
+    response.sendRedirect("/");
+  }
+
+  /**
+   * Add the username to the database, or replace the current username.
+   */
+  public static void putUsername(String username, String id) {
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    Entity entity = new Entity("UserInfo", id);
+    entity.setProperty("id", id);
+    entity.setProperty("username", username);
+    // The put() function automatically inserts new data or updates existing 
+    // data based on ID
+    datastore.put(entity);
+  }
+
+  /**
+   * @return the request parameter, or the default value if the parameter
+   *         was not specified by the client
+   */
+  private String getParameter(HttpServletRequest request, String name, String defaultValue) {
+    String value = request.getParameter(name);
+    if (value == null) {
+      return defaultValue;
+    }
+    return value;
+  }
+
+  /** 
+   * Returns the username of the user with id, or null if the user has not
+   * set a username.
```

Comment:
I think this comment is out of date. The method below will return (and store) the email prefix if the user hasn't directly set a username.

Reply:
Done!